File: src/main/java/com/github/joelws/release/tracker/factory/ReleaseTrackerFactory.java
package com.github.joelws.release.tracker.factory;

import com.github.joelws.release.tracker.exception.ReleaseTrackerException;
import com.github.joelws.release.tracker.handler.JsonResponse;

import org.apache.log4j.Logger;

public class ReleaseTrackerFactory implements Factory
{
    private final static Logger LOGGER = Logger.getLogger(ReleaseTrackerFactory.class);

    @Override public <T> T getImpl(Class<T> klazz)
    {
        try
        {
            return klazz.newInstance();
        }
        catch (InstantiationException | IllegalAccessException e)
        {
            LOGGER.error("Failed to instantiate object: " + klazz);
            throw new ReleaseTrackerException(new JsonResponse(500, "Failed to instantiate object"));
        }
    }
}

File: src/main/java/com/github/joelws/release/tracker/factory/ReleaseTrackerFactory.kt
package com.github.joelws.release.tracker.factory

import com.github.joelws.release.tracker.exception.ReleaseTrackerException
import com.github.joelws.release.tracker.response.RestResponse
import org.apache.log4j.Logger

open class ReleaseTrackerFactory : Factory {

    companion object {
        val LOGGER = Logger.getLogger(ReleaseTrackerFactory::class.java)
    }

    override fun <T> getImpl(klazz: Class<T>) = try {
        klazz.newInstance()
    } catch (e: InstantiationException) {
        LOGGER.error("Failed to instantiate object: " + klazz)
        throw ReleaseTrackerException(RestResponse.ServerError("Failed to instantiate object"))
    } catch(e: IllegalAccessException) {
        LOGGER.error("Failed to instantiate object: " + klazz)
        throw ReleaseTrackerException(RestResponse.ServerError("Failed to instantiate object"))
    }
}
